core/services/pipeline/orm.go

```diff
@@ -51,6 +58,145 @@ func (o *orm) CreateSpec(ctx context.Context, tx *gorm.DB, pipeline Pipeline, ma
 	return spec.ID, errors.WithStack(err)
 }
 
+// InsertRun
+func (o *orm) CreateRun(db *gorm.DB, run *Run) (err error) {
+	if run.CreatedAt.IsZero() {
+		return errors.New("run.CreatedAt must be set")
+	}
+	if err = db.Create(run).Error; err != nil {
+		return errors.Wrap(err, "error inserting finished pipeline_run")
+	}
+	return err
+}
+
+// StoreRun
+func (o *orm) StoreRun(db *sql.DB, run *Run, saveSuccessfulTaskRuns bool) (bool, error) {
+	finished := run.FinishedAt != nil
+	var restart bool
+	err := postgres.SqlxTransaction(context.Background(), db, func(tx *sqlx.Tx) error {
+		if !finished {
+			// Lock the current run. This prevents races with /v2/resume
+			sql := `SELECT id FROM pipeline_runs WHERE id = $1 FOR UPDATE;`
```

Comment:
The state change `running -> suspended` only happens if no new task run data is detected. If new data is present, we keep the state as running, then return a restart flag to resume execution.

Reply:
I guess we could eagerly update to `suspended`, since in the majority of cases we don't expect `resume` to be called so early? That would make the happy path use one query less, but `if restart` would then need a second query to update back to `running`.

Reply:
Makes sense, this current implementation seems fine then